### PR TITLE
Fix for issue #53 - Receiving "Address Already in Use" error when binding two sockets to same port on iOS

### PR DESCRIPTION
**Pull request by intmainvoid**
1. Updated GCDAsyncUdpSocket
2. Set option to reuse port if socket bound to same port twice from same IP Address

### CODE REVIEW

**Comment by intmainvoid** [comment]
Fix for issue #53

**Comment by mvayngrib** [comment]
@intmainvoid thanks for the PR! I haven't reviewed the changes yet, but meanwhile, to be consistent with dgram's behavior, I think you should only allow reuse when the `createSocket` is called with `reuseAddr: true`:

see:
- https://nodejs.org/api/dgram.html#dgram_dgram_createsocket_options_callback
- https://github.com/nodejs/node/issues/2604

also:
do you know if this GCDAsyncUdpSocket introduces any breaking changes, e.g. what's the semver change in GCDAsyncUdpSocket?

**Comment by intmainvoid** [comment]
> I think you should only allow reuse when the createSocket is called with reuseAddr: true:

reuseAddr is quite different to reusePort on the POSIX socket level. (See [here](https://stackoverflow.com/questions/14388706/socket-options-so-reuseaddr-and-so-reuseport-how-do-they-differ-do-they-mean-t)). Anyway I've updated the code and tried to make it adhere to the pattern as closely as possible to the patterns used in node.js's dgram. However, it seems that dgram doesn't actually use the SO_REUSEPORT option.

> do you know if this version of GCDAsyncUdpSocket introduces any breaking changes, e.g. what's the semver change in GCDAsyncUdpSocket?

Tested relatively thoroughly in a new iOS app that is soon to be released. Crashes went away after this change (although I suspect it is due more to the introduction of the reusePort option than the update of GCDAsyncUdpSocket).

**Comment by intmainvoid** [comment]
Oh just read: nodejs/node#2604. Based on that, node.js turns on SO_REUSEPORT without any option for the user to disable/enable it for OSX. This is how my original PR handled this. In any case, I'm happy either way as long as there is a way to turn it on.

**Comment by mvayngrib** [comment]
@intmainvoid it's not entirely clear to me how reuseAddr works in node, but given this snippet (adapted from nodejs/node#2604), if you run it in two terminal tabs, it only works when reuseAddr is true:

```js
const dgram = require('dgram')

dgram.createSocket({
  type: 'udp4',
  reuseAddr: true
})
.bind({
  address: '127.0.0.1',
  port: 9527
})
```

**Comment by intmainvoid** [comment]
Yep, that all sounds good. So this PR proposes that this feature be used in the following way:
```
const socket = dgram.createSocket({
  type: 'udp4',
  reusePort: true
}).bind('127.0.0.1', (err) => {
  const success = !err;
});
```
I think this change to the react-native-udp's API follows your example for reuseAddr very closely.

**Comment by quhaoran007** [comment]
Hi @mvayngrib, please follow up this. This issue happens very often to my app when another app is also using UDP and being kill by OS in the background(you can see from the multi-task screen). I think somehow OS wakes up the other app and that app binds the port, so my app using this library can't bind the port.  Have used "enableReusePort", but it didn't help.

**Comment by quhaoran007** [comment]
Hi @mvayngrib, please follow up this. This issue happens very often to my app when another app is also using UDP and being kill by OS in the background(you can see from the multi-task screen). I think somehow OS wakes up the other app and that app binds the port, so my app using this library can't bind the port.  Have used "enableReusePort", but it didn't help.

**Comment by mvayngrib** [comment]
@intmainvoid sorry, missed your update of the PR. Can you tell me to what version of GCDAsyncUdpSocket you updated? It would be good to have the commit referenced from https://github.com/robbiehanson/CocoaAsyncSocket

thanks!

File: UdpSocket.js [hunk 1]
```diff
@@ -112,7 +112,7 @@ UdpSocket.prototype.bind = function(port, address, callback) {
 
 UdpSocket.prototype.close = function (callback=noop) {
   if (this._destroyed) {
-    return process.nextTick(callback)
+    return setImmediate(callback)
```
Comment (mvayngrib) [inline UdpSocket.js]: any particular reason?

Comment (intmainvoid) [reply UdpSocket.js]: process.nextTick doesn't exist on React Native. This line was generating an error

Comment (mvayngrib) [reply UdpSocket.js]: ah, right, i've gotten so used to using rn-nodeify, i've forgotten i'm not in node. Good catch!

**Comment by mvayngrib** [comment]
for future reference, the current version (2.1.0) uses CocoaAsyncSocket 7.4.1, commit [c0bbcbc](https://github.com/robbiehanson/CocoaAsyncSocket/commit/c0bbcbcc5e039ca5d732f9844bf95c3d8ee31a5b)

**Comment by intmainvoid** [comment]
> @intmainvoid sorry, missed your update of the PR. Can you tell me to what version of GCDAsyncUdpSocket you updated? It would be good to have the commit hash from https://github.com/robbiehanson/CocoaAsyncSocket

I believe it is this one:
[d0adf58ca694e733c75a8a157635e3deb66c061e](https://github.com/robbiehanson/CocoaAsyncSocket/commit/d0adf58ca694e733c75a8a157635e3deb66c061e)

So, basically the current master branch.

**Comment by mvayngrib** [comment]
cool, that checks out :)

**Comment by mvayngrib** [comment]
@intmainvoid released as 2.2.0, thanks again for the PR!
@quhaoran007 thanks for the poke :)

**Pull request merged**